### PR TITLE
Proxmox inventory: implement API token auth

### PR DESCRIPTION
**Pull request by h3po**
##### SUMMARY
Adds api token authentication as an alternative to password auth to the proxmox inventory plugin.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
proxmox

##### ADDITIONAL INFORMATION
I'm unsure what to set version_added to for the new environment variables; also how to correctly raise an error if neither password nor token_id + token_secret are specified (see Line 595)

### CODE REVIEW

File: plugins/inventory/proxmox.py [hunk 1]
```diff
@@ -52,11 +52,33 @@
...
+        required: no
         type: str
         env:
           - name: PROXMOX_PASSWORD
             version_added: 2.0.0
+      token_id:
+        description:
+          - Proxmox authentication token id.
+          - If the value is not specified in the inventory configuration, the value of environment variable C(PROXMOX_TOKEN_ID) will be used instead.
+          - To use token authentication, you must also specify token_secret. If you do not specify token_id and token_secret, you must set a password instead.
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:65]:
```suggestion
          - To use token authentication, you must also specify I(token_secret). If you do not specify I(token_id) and I(token_secret), you must set a password instead.
```

File: plugins/inventory/proxmox.py [hunk 2]
```diff
@@ -52,11 +52,33 @@
...
+          - Make sure to grant explicit pve permissions to the token or disable 'privilege separation' to use the users' privileges instead
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:66]:
```suggestion
          - Make sure to grant explicit pve permissions to the token or disable 'privilege separation' to use the users' privileges instead.
```

File: plugins/inventory/proxmox.py [hunk 3]
```diff
@@ -52,11 +52,33 @@
...
+        type: str
+        env:
+          - name: PROXMOX_TOKEN_ID
+            version_added: 4.7.1
+      token_secret:
+        description:
+          - Proxmox authentication token secret.
+          - If the value is not specified in the inventory configuration, the value of environment variable C(PROXMOX_TOKEN_SECRET) will be used instead.
+          - To use token authenticaiton, you must also specify token_id. If you do not specify token_id and token_secret, you must set a password instead.
+        required: no
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:77]:
```suggestion
        version_added: 4.8.0
```

File: plugins/inventory/proxmox.py [hunk 4]
```diff
@@ -52,11 +52,33 @@
...
+        type: str
+        env:
+          - name: PROXMOX_TOKEN_ID
+            version_added: 4.7.1
+      token_secret:
+        description:
+          - Proxmox authentication token secret.
+          - If the value is not specified in the inventory configuration, the value of environment variable C(PROXMOX_TOKEN_SECRET) will be used instead.
+          - To use token authenticaiton, you must also specify token_id. If you do not specify token_id and token_secret, you must set a password instead.
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:76]: Same changes as above.

File: plugins/inventory/proxmox.py [hunk 5]
```diff
@@ -222,15 +244,24 @@ def _get_session(self):
     def _get_auth(self):
         credentials = urlencode({'username': self.proxmox_user, 'password': self.proxmox_password, })
 
-        a = self._get_session()
-        ret = a.post('%s/api2/json/access/ticket' % self.proxmox_url, data=credentials)
+        if (self.proxmox_password):
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:247]: This is Python, not C :-)
```suggestion
        if self.proxmox_password:
```

File: plugins/inventory/proxmox.py [hunk 6]
```diff
@@ -552,6 +582,19 @@ def parse(self, inventory, loader, path, cache=True):
...
+        if t.is_template(proxmox_token_id):
+            proxmox_token_id = t.template(variable=proxmox_token_id, disable_lookups=False)
+        self.proxmox_token_id = proxmox_token_id
+
+        proxmox_token_secret = self.get_option('token_secret')
+        if t.is_template(proxmox_token_secret):
+            proxmox_token_secret = t.template(variable=proxmox_token_secret, disable_lookups=False)
+        self.proxmox_token_secret = proxmox_token_secret
+
+        if not (proxmox_password != None) and not (proxmox_token_id != None and proxmox_token_secret != None):
```
Comment (felixfontein) [inline plugins/inventory/proxmox.py:595]:
```suggestion
        if proxmox_password is None and (proxmox_token_id is None or proxmox_token_secret is None):
```

**Review by felixfontein — COMMENTED** [review]
Thanks for your contribution! I've added some first comments. Could you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment)? Thanks.

**Comment by h3po** [comment]
@felixfontein thanks for your review, it should all be in order now.

File: changelogs/fragments/4540-proxmox-inventory-token-auth.yml [hunk 7]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - proxmox inventory plugin - add token authentication as an alternative to username/password.
```
Comment (felixfontein) [inline changelogs/fragments/4540-proxmox-inventory-token-auth.yml:2]:
```suggestion
  - proxmox inventory plugin - add token authentication as an alternative to username/password (https://github.com/ansible-collections/community.general/pull/4540).
```

**Comment by felixfontein** [comment]
@Ajpantuso @Thulium-Drake @ilijamt @joshainglis @karmab @tleguern would be glad if someone of you could take a look at this!

**Review by felixfontein — APPROVED** [review]

**Comment by Thulium-Drake** [comment]
Looks nice! I haven't tested the changes, I've been a bit pre-occupied last few weeks.

Thanks for your work!

**Comment by felixfontein** [comment]
@h3po thanks for implementing this!
@Thulium-Drake thanks for reviewing!

**Pull request merged**